src/encoder/encoder.h

```diff
@@ -17,23 +18,56 @@
 #ifndef ENCODER_H
 #define ENCODER_H
 
+#include "util/memory.h"
 #include "util/types.h"
+#include "preferences/usersettings.h"
+#include "encoder/encodersettings.h"
+#include "encoder/encodercallback.h"
 
 class Encoder {
   public:
-    Encoder();
-    virtual ~Encoder();
+        class Format {
+            public:
+            Format(QString labelIn, QString nameIn, bool losslessIn) :
+                label(labelIn), internalName(nameIn), lossless(losslessIn) {}
+            QString label;
+            QString internalName;
+            bool lossless;
+        };
 
-    virtual int initEncoder(int bitrate, int samplerate) = 0;
+    Encoder() {}
+    virtual ~Encoder() {}
+
+    virtual int initEncoder(int samplerate, QString errorMessage) = 0;
     // encodes the provided buffer of audio.
     virtual void encodeBuffer(const CSAMPLE *samples, const int size) = 0;
-    // Adds metadata to the encoded auio, i.e., the ID3 tag. Currently only used
+    // Adds metadata to the encoded audio, i.e., the ID3 tag. Currently only used
     // by EngineRecord, EngineBroadcast does something different.
-    virtual void updateMetaData(char* artist, char* title, char* album) = 0;
+    virtual void updateMetaData(const char* artist, const char* title, const char* album) = 0;
     // called at the end when encoding is finished
     virtual void flush() = 0;
-    /**converts an OGG quality measure from 1..10 to a bitrate **/
-    static int convertToBitrate(int quality);
+    // Setup the encoder with the specific settings
+    virtual void setEncoderSettings(const EncoderSettings& settings) = 0;
+};
+
+typedef std::shared_ptr<Encoder> EncoderPointer;
```

Comment:
I'm a bit new to shared/unique pointers. I used shared_ptr because I wanted to have the automatic deletion of objects when the pointer is discarded but I don't fully know all the details. So.. since I return it on a method that always creates a new pointer, it should be an unique_ptr?